### ISSUE TITLE
[Platform] RBAC UI updates

### ISSUE THREAD

**Issue by ramkumarvs** [yugabyte/yugabyte-db#3794]
We need to handle couple of things on our Platform UI 
- Read Only user:
  - We need to hide software upgrade and other stuff
  - Hide the cost for read only user
  - Disable snooze, health, backup buttons etc.
  - Delete user for a readonly user should be disabled
- Admin user:
 - Disable delete of superadmin

Aha! Link: https://yugabyte-test.aha.io/features/PLATFORM-653

**Comment by kriteshranjan** [comment]
AI validation notes (`yb-issue-3794-analysis.md`):

# Analysis: yugabyte/yugabyte-db#3794 — [Platform] RBAC UI updates

- **Filed:** 2020-02-28 · **State:** open · **Labels:** area/platform · **Comments:** 0
- **Kind:** feature request (YBA management-console UI/RBAC gating) → behavior-delta validation, not a crash repro.
- **Requested items:** ReadOnly user: (1) hide software upgrade, (2) hide cost, (3) disable snooze/health/backup buttons, (4) disable delete-user; Admin user: (5) disable delete of SuperAdmin.
- **Validated:** 2026-08-29 against local master checkout `46a9f7adfd` (2026-08-28).

## Verdict
NOT-REPRODUCED — the requested gap no longer exists on current master: all five requested role-gating behaviors are implemented (four of them since 2020-03-12 by issue-tagged commit `f8456a530c53`; the SuperAdmin-delete hard guard completed 2026-06-01 by `c67f31b59eb8`), so the issue is fully delivered and can be closed as fixed.

## Root cause
The 2020 gap and its delivery are both located in `managed/` (YBA is open-source in-tree). The delivery mechanism is a per-role feature-config overlay plus, later, fine-grained per-API RBAC:

- `managed/src/main/java/com/yugabyte/yw/models/Users.java:101-116` — `Role.getFeaturesFile()` maps `ReadOnly` → `readOnlyFeatureConfig.json` (line 105-106).
- `managed/src/main/java/com/yugabyte/yw/common/user/UserService.java:36-47` — loads that JSON into `UserWithFeatures`.
- `managed/src/main/java/com/yugabyte/yw/controllers/TokenAuthenticator.java:238` — attaches `UserWithFeatures` to every authenticated request.
- `managed/src/main/java/com/yugabyte/yw/controllers/CustomerController.java:169-177` — `GET /customers/:cUUID` deep-merges the user's role features into the customer features the UI consumes as `currentCustomer.data.features`.
- UI helpers (`isDisabled` / `isNotHidden` / `isAvailable` / `getFeatureState` from `managed/ui/src/utils/LayoutUtils.js`) read those keys at each gated control (call sites in the Trigger matrix below).
- Backend hard guard for item (5): `managed/src/main/java/com/yugabyte/yw/controllers/UsersController.java:297-306` — `delete()` rejects with 400 `"Cannot delete a SuperAdmin user ..."` unless the caller passes `superAdminAuthentication`. Companion guards: `UsersController.java:193-194` (cannot create a SuperAdmin), `UsersController.java:395-396` (cannot change a role to SuperAdmin).

Master is not affected by the reported gap: nothing remains to fix. The legacy role model is additionally superseded by the fine-grained RBAC framework (`managed/src/main/java/com/yugabyte/yw/common/rbac/`, `managed/ui/src/redesign/features/rbac/`) with per-API permission gating (`RbacValidator` + `ApiPermissionMap`), enabled by default 2024-11-25 (`8622f0abc82f`, PLAT-14769) and GA in 2025.1.

## Trigger matrix
Behavior-delta matrix (feature request): predicted delivery state from commit lineage vs observed state in master source. All five cells observed as predicted.

| # | Requested (2020-02-28) | Predicted (from lineage) | Observed on master `46a9f7adfd` |
|---|---|---|---|
| 1 | ReadOnly: hide software upgrade | Delivered (pre-dated fix commit; parent `a193dd6bb132` already had `upgradeSoftware: "hidden"`) | `readOnlyFeatureConfig.json:33` `"upgradeSoftware": "hidden"`; consumed at `managed/ui/src/components/universes/UniverseDetail/UniverseDetail.jsx:1108-1124` ("Upgrade Software" menu item availability) and 3 more sites (`:1143,:1168,:1200`); each also wrapped in `RbacValidator` with `MODIFY_UNIVERSE` |
| 2 | ReadOnly: hide cost | Delivered by `f8456a530c53` (adds `costs`/`main.stats` hidden) | `readOnlyFeatureConfig.json:41` `"costs": "hidden"` → `managed/ui/src/components/universes/UniverseOverview/UniverseOverviewNew.jsx:814`; `readOnlyFeatureConfig.json:52-54` `"main": {"stats": "hidden"}` → `managed/ui/src/components/dashboard/Dashboard.jsx:39` |
| 3 | ReadOnly: disable snooze/health/backup buttons | Delivered by `f8456a530c53` (adds `universes.actions`/`universes.backup` disabled; wires health/backup buttons) | `readOnlyFeatureConfig.json:45-47` (`tableActions`/`actions`/`backup` disabled), `:13-15` (`health.configure` disabled) → `ListTables.jsx:224,380`, `Schedules.jsx:217,237`, `NodeDetailsTable.jsx:281`, `UniverseDetail.jsx:1864`, `AlertProfileForm.jsx:104`; the modern snooze control ("Create Maintenance Window", `managed/ui/src/components/universes/UniverseAction/UniverseAction.jsx:16-36`) is permission-gated via `RbacValidator` on `CREATE_MAINTENANCE_WINDOW` |
| 4 | ReadOnly: disable delete-user | Delivered by `f8456a530c53` (UserList gated on `universe.create` disabled) | Legacy list: `managed/ui/src/components/users/Users/UsersList.jsx:66-73` — Actions dropdown (incl. Delete User) rendered as `null` when `isDisabled(customer.data.features, 'universe.create')` (true for ReadOnly per `readOnlyFeatureConfig.json:4`); new RBAC UI: `managed/ui/src/redesign/features/rbac/users/components/ViewUsers.tsx:240-250` requires `DELETE_USER` permission on the user resource |
| 5 | Admin: disable delete of SuperAdmin | Delivered: UI suppression (legacy + RBAC UI), backend hard guard by `c67f31b59eb8` (2026-06-01) | Backend: `UsersController.java:299-306` (400 unless caller is SuperAdmin), regression test `UsersControllerTest.testDeleteSuperAdminUserAsAdminFails`; UI: `UsersList.jsx:71` (`user.role === 'SuperAdmin'` → no Actions dropdown), `ViewUsers.tsx:252` (`disabled: isSuperAdmin || self`) |

## Reproduction
Executable reproduction does not apply: this is a feature request against the YBA management console's role-gated UI, not a database-daemon defect — exercising it would require a deployed multi-user YBA instance with browser interaction, and there is no failure signature to grep for. Per the validation protocol, feature requests get behavior-delta validation (current vs requested), performed above against the in-tree open-source `managed/` code. The validation is deterministic and rerunnable from any yugabyte-db checkout:

```sh
# Delivering commits (GitHub commit search; local checkout is shallow):
gh search commits --repo yugabyte/yugabyte-db "3794"
#  -> f8456a530c53 2020-03-12 "[#3794] Add support for disabling readonly users"
gh search commits --repo yugabyte/yugabyte-db "Cannot delete a SuperAdmin"
#  -> c67f31b59eb8 2026-06-01 "[PLAT-21099] YBA: Restrict user deletion ..."

# Current-state checks (run at repo root; expect matches at the cited lines):
cat managed/src/main/resources/readOnlyFeatureConfig.json
grep -n "universes.details.overview.upgradeSoftware" \
  managed/ui/src/components/universes/UniverseDetail/UniverseDetail.jsx
grep -n "universes.details.overview.costs" \
  managed/ui/src/components/universes/UniverseOverview/UniverseOverviewNew.jsx
grep -n "main.stats" managed/ui/src/components/dashboard/Dashboard.jsx
sed -n "66,73p" managed/ui/src/components/users/Users/UsersList.jsx
sed -n "297,306p" managed/src/main/java/com/yugabyte/yw/controllers/UsersController.java
sed -n "252p" managed/ui/src/redesign/features/rbac/users/components/ViewUsers.tsx
```

Full captured command outputs: `/tmp/claude/yb-3794-rbac-evidence.txt`. Diff of the issue-tagged delivering commit: `/tmp/claude/yb-3794-commit-f8456a530c53.diff`.

## Versions tested
- Source state verified on master `46a9f7adfd226d7bc3315c38b660eef69bb11196` (2026-08-28). Master affected: **no** (all requested behaviors present).
- Lineage: ReadOnly items shipped with `f8456a530c53` (2020-03-12, Phab D8109 — 2.1.x-era YBA, ~13 days after the issue was filed) atop the pre-existing ReadOnly feature-config (`a193dd6bb132` already hid `upgradeSoftware`); SuperAdmin-delete backend hard guard shipped with `c67f31b59eb8` (2026-06-01, PLAT-21099, Phorge D53875), preceded by UI-level suppression and an earlier "cannot delete primary user" guard; fine-grained RBAC on-by-default since `8622f0abc82f` (2024-11-25), GA 2025.1.
- No container/binary run was applicable (UI/management-plane feature request; behavior-delta validation only).

## Proposed fix direction
No code fix needed. Recommended action: close #3794 as completed, citing `f8456a530c53` (issue-tagged, delivered the ReadOnly items in 2020) and `c67f31b59eb8` (completed the SuperAdmin-delete hard guard), noting supersession by the fine-grained RBAC framework (`managed/ui/src/redesign/features/rbac/`, `managed/src/main/java/com/yugabyte/yw/common/rbac/`) which replaces the legacy 5-role gating with per-API, per-resource permissions.

## Test strategy
Not applicable for a fix (none needed). Existing coverage guarding the delivered behavior: `managed/src/test/java/com/yugabyte/yw/controllers/UsersControllerTest.java` — `testDeleteSuperAdminUserAsAdminFails` (added by `c67f31b59eb8`) asserts an Admin deleting a SuperAdmin receives 400 and the user persists; run with `cd managed && sbt "testOnly com.yugabyte.yw.controllers.UsersControllerTest"`. UI feature-config gating is exercised by YBA UI tests (`managed/ui`, `npm run test:ci`).

## Confidentiality check
Contains only public facts: public GitHub issue text, public commit SHAs/messages (which themselves carry the Phab/Phorge revision IDs quoted here), and in-tree open-source file:line citations. No customer data, PII, internal hostnames/URLs, or secrets.


[PLAT-21099]: https://yugabyte.atlassian.net/browse/PLAT-21099?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

**Comment by kriteshranjan** [comment]
AI validation notes (`yb-3794-rbac-evidence.txt`):

Evidence: yugabyte/yugabyte-db#3794 "[Platform] RBAC UI updates" (filed 2020-02-28, open)
Validation date: 2026-08-29. Local checkout: master 46a9f7adfd (2026-08-28).
All commands rerunnable from a yugabyte-db checkout root (gh = GitHub CLI).
================================================================================

--- [A] Delivering commit for the ReadOnly-user items (issue-tagged) ---
$ gh search commits --repo yugabyte/yugabyte-db "3794"
f8456a530c53 2020-03-12 [#3794] Add support for disabling readonly users
55c219b73fcb 2022-06-01 [PLAT-3794] Deletion of read clusters for Kubernetes universes.

f8456a530c53 (2020-03-12, Phab D8109) "[#3794] Add support for disabling readonly users"
"Summary: Make sure readonly user has all the buttons disabled or hidden"
Files: readOnlyFeatureConfig.json + 12 UI components (UserList, UserProfileForm,
ReleaseList, ListBackups, ListTables, TableAction, NodeAction, NodeDetailsTable,
UniverseAction, UniverseDetail, UniverseHealthCheckList, UniverseOverviewNew).
Key hunks: adds "costs": "hidden", "actions": "disabled", "backup": "disabled",
"main": { "stats": "hidden" } to readOnlyFeatureConfig.json; gates user-list action
buttons on isDisabled(features, "universe.create").
(NOTE: "upgradeSoftware": "hidden" already present in parent commit a193dd6bb132 —
software-upgrade hiding for ReadOnly pre-dated this issue's fix commit.)

--- [B] Delivering commit for the Admin-user item (SuperAdmin delete guard) ---
$ gh search commits --repo yugabyte/yugabyte-db "Cannot delete a SuperAdmin"
c67f31b59eb8 2026-06-01 [PLAT-21099] YBA: Restrict user deletion — block Admin from deleting SuperAdmin and prevent SuperAdmin self-delete
c67f31b59eb8 (2026-06-01, PLAT-21099, Phorge D53875): backend role-based guard replacing
the earlier blanket "cannot delete primary user" check; adds
UsersControllerTest.testDeleteSuperAdminUserAsAdminFails (Admin deleting SuperAdmin -> 400).

--- [C] Current master: ReadOnly role feature config (verbatim) ---
$ cat managed/src/main/resources/readOnlyFeatureConfig.json
{
  "universe": {
    "import": "disabled",
    "create": "disabled"
  },
  "config": {
    "infra": "disabled",
    "backup": "disabled"
  },
  "menu": {
    "config": "disabled"
  },
  "health": {
    "configure": "disabled"
  },
  "alert": {
    "list": {
      "actions": "disabled"
    },
    "configuration": {
      "actions": "disabled"
    },
    "destinations": {
      "actions": "disabled"
    },
    "channels": {
      "actions": "disabled"
    }
  },
  "universes": {
    "details": {
      "overview": {
        "upgradeSoftware": "hidden",
        "editGFlags": "hidden",
        "editUniverse": "hidden",
        "readReplica": "hidden",
        "manageEncryption": "hidden",
        "restartUniverse": "hidden",
        "pausedUniverse": "disabled",
        "deleteUniverse": "hidden",
        "costs": "hidden",
        "systemdUpgrade": "hidden"
      }
    },
    "tableActions": "disabled",
    "actions": "disabled",
    "backup": "disabled"
  },
  "administration": {
    "highAvailability": "hidden"
  },
  "main": {
    "stats": "hidden"
  },
  "profile": {
    "profileInfo": "disabled"
  }
}

--- [D] Current master: mechanism chain (role -> features -> UI) ---
$ grep -n "readOnlyFeatureConfig\|getFeaturesFile" managed/src/main/java/com/yugabyte/yw/models/Users.java
101:    public String getFeaturesFile() {
106:          return "readOnlyFeatureConfig.json";

$ sed -n "36,47p" managed/src/main/java/com/yugabyte/yw/common/user/UserService.java
  public UserWithFeatures getUserWithFeatures(Customer customer, Users user) {
    try {
      UserWithFeatures userWithFeatures = new UserWithFeatures().setUser(user);
      String configFile = user.getRole().getFeaturesFile();
      if (configFile == null) {
        return userWithFeatures;
      }
      try (InputStream featureStream = environment.resourceAsStream(configFile)) {
        ObjectMapper mapper = new ObjectMapper();
        JsonNode features = mapper.readTree(featureStream);
        return userWithFeatures.setFeatures(features);
      }

$ grep -n "getUserWithFeatures" managed/src/main/java/com/yugabyte/yw/controllers/TokenAuthenticator.java
238:        RequestContext.put(USER, userService.getUserWithFeatures(cust, user));

$ sed -n "169,177p" managed/src/main/java/com/yugabyte/yw/controllers/CustomerController.java
    UserWithFeatures user = RequestContext.get(TokenAuthenticator.USER);
    if (customer.getFeatures().size() != 0 && user.getFeatures().size() != 0) {
      JsonNode featureSet = user.getFeatures();
      CommonUtils.deepMerge(featureSet, customer.getFeatures());
      responseJson.put("features", featureSet);
    } else if (customer.getFeatures().size() != 0) {
      responseJson.put("features", customer.getFeatures());
    } else {
      responseJson.put("features", user.getFeatures());

--- [E] Current master: per-sub-item UI gating call sites ---
# (1) hide software upgrade for ReadOnly:
$ grep -n "universes.details.overview.upgradeSoftware" managed/ui/src/components/universes/UniverseDetail/UniverseDetail.jsx
1113:                            'universes.details.overview.upgradeSoftware'
1143:                            'universes.details.overview.upgradeSoftware'
1168:                            'universes.details.overview.upgradeSoftware'
1200:                              'universes.details.overview.upgradeSoftware'

# (2) hide cost for ReadOnly:
$ grep -n "universes.details.overview.costs" managed/ui/src/components/universes/UniverseOverview/UniverseOverviewNew.jsx
814:          {isEnabled(currentCustomer.data.features, 'universes.details.overview.costs') &&
$ grep -n "main.stats" managed/ui/src/components/dashboard/Dashboard.jsx
39:        {isAvailable(currentCustomer.data.features, 'main.stats') && (

# (3) disable snooze/health/backup buttons for ReadOnly:
$ grep -rn "universes.backup\|universes.actions" managed/ui/src/components --include="*.jsx" | grep -v test
managed/ui/src/components/tables/ListTables/ListTables.jsx:224:      isDisabled(currentCustomer.data.features, 'universes.backup') ||
managed/ui/src/components/tables/ListTables/ListTables.jsx:380:          isNotHidden(currentCustomer.data.features, 'universes.backup') &&
managed/ui/src/components/schedules/Schedules.jsx:217:                readOnly={isNonAvailable(currentCustomer.data.features, 'universes.backup')}
managed/ui/src/components/schedules/Schedules.jsx:237:                {isAvailable(currentCustomer.data.features, 'universes.backup') && (
managed/ui/src/components/universes/UniverseDetail/UniverseDetail.jsx:1864:                              'universes.backup'
managed/ui/src/components/universes/NodeDetails/NodeDetailsTable.jsx:281:        'universes.actions'
$ grep -n "health.configure" managed/ui/src/components/profile/AlertProfileForm.jsx
104:    const isReadOnly = isDisabled(customer.data.features, 'health.configure');

# (4) delete-user disabled for ReadOnly actor (legacy list: dropdown suppressed):
$ sed -n "66,73p" managed/ui/src/components/users/Users/UsersList.jsx
  const actionsDropdown = (user) => {
    const loginUserId = localStorage.getItem('userId');
    if (
      user.uuid === loginUserId ||
      isDisabled(customer.data.features, 'universe.create') ||
      user.role === 'SuperAdmin'
    ) {
      return null;

# (5) SuperAdmin delete protection (backend hard guard):
$ sed -n "297,306p" managed/src/main/java/com/yugabyte/yw/controllers/UsersController.java
  public Result delete(UUID customerUUID, UUID userUUID, Http.Request request) {
    Users user = Users.getOrBadRequest(customerUUID, userUUID);
    if (userHasSuperAdminRole(user, customerUUID)
        && !tokenAuthenticator.superAdminAuthentication(request)) {
      throw new PlatformServiceException(
          BAD_REQUEST,
          String.format(
              "Cannot delete a SuperAdmin user %s for customer %s",
              userUUID.toString(), customerUUID));
    }

# (5) SuperAdmin delete protection (new RBAC UI: delete menu item):
$ sed -n "252p" managed/ui/src/redesign/features/rbac/users/components/ViewUsers.tsx
            disabled: isSuperAdmin || user.uuid === localStorage.getItem('userId')

--- [F] Supersession: fine-grained RBAC framework (per-API permission gating) ---
$ ls managed/src/main/java/com/yugabyte/yw/common/rbac/ managed/ui/src/redesign/features/rbac/ | head -20
Permission.java
PermissionInfo.java
PermissionUtil.java
RoleBindingUtil.java
RoleResourceDefinition.java
RoleUtil.java
routes
api.ts
ApiAndUserPermMapping.ts
common
groups
index.ts
permission
policy
RBACContainer.tsx
$ gh search commits --repo yugabyte/yugabyte-db "fine-grained RBAC" (oldest relevant)
8622f0abc82f 2024-11-25 [PLAT-14769]Enable fine grained RBAC by default
4b6e1d994a43 2025-06-26 [DOC-818][yba][2025.1] Fine-grain RBAC GA (#27724)


[PLAT-3794]: https://yugabyte.atlassian.net/browse/PLAT-3794?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[PLAT-21099]: https://yugabyte.atlassian.net/browse/PLAT-21099?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

**Comment by kriteshranjan** [comment]
Closing this issue based on AI validation. cc: @m-iancu , @VamsidharGunturu
